cl/phase1/network/services/blob_sidecar_service.go

```diff
@@ -0,0 +1,154 @@
+	b := &blobSidecarService{
+		beaconCfg:         beaconCfg,
+		forkchoiceStore:   forkchoiceStore,
+		syncedDataManager: syncedDataManager,
+	}
+	go b.loop(ctx)
+	return b
+}
+
+// ProcessMessage processes a blob sidecar message
+func (b *blobSidecarService) ProcessMessage(ctx context.Context, msg *cltypes.BlobSidecar) error {
+	headState := b.syncedDataManager.HeadState()
+	if headState == nil {
+		b.scheduleBlobSidecarForLaterExecution(msg)
+		return ErrIgnore
+	}
+
+	parentHeader, has := b.forkchoiceStore.GetHeader(msg.SignedBlockHeader.Header.ParentRoot)
+	if !has {
+		b.scheduleBlobSidecarForLaterExecution(msg)
+		return ErrIgnore
+	}
+	if msg.SignedBlockHeader.Header.Slot <= parentHeader.Slot {
+		return ErrInvalidSidecarSlot
+	}
+	return b.verifyAndStoreBlobSidecar(headState, msg)
+}
+
+func (b *blobSidecarService) verifyAndStoreBlobSidecar(headState *state.CachingBeaconState, msg *cltypes.BlobSidecar) error {
+	kzgCtx := kzg.Ctx()
+
+	if !cltypes.VerifyCommitmentInclusionProof(msg.KzgCommitment, msg.CommitmentInclusionProof, msg.Index,
+		clparams.DenebVersion, msg.SignedBlockHeader.Header.BodyRoot) {
+		return ErrCommitmentsInclusionProofFailed
+	}
+
+	if err := kzgCtx.VerifyBlobKZGProof(gokzg4844.Blob(msg.Blob), gokzg4844.KZGCommitment(msg.KzgCommitment), gokzg4844.KZGProof(msg.KzgProof)); err != nil {
+		return fmt.Errorf("blob KZG proof verification failed: %v", err)
+	}
+
+	if err := b.verifySidecarsSignature(headState, msg.SignedBlockHeader); err != nil {
+		return err
+	}
+	// operation is not thread safe from here.
+	return b.forkchoiceStore.AddPreverifiedBlobSidecar(msg)
+}
+
+func (b *blobSidecarService) verifySidecarsSignature(headState *state.CachingBeaconState, header *cltypes.SignedBeaconBlockHeader) error {
+	parentHeader, ok := b.forkchoiceStore.GetHeader(header.Header.ParentRoot)
+	if !ok {
+		return fmt.Errorf("parent header not found")
+	}
+	currentVersion := b.beaconCfg.GetCurrentStateVersion(parentHeader.Slot / b.beaconCfg.SlotsPerEpoch)
+	forkVersion := b.beaconCfg.GetForkVersionByVersion(currentVersion)
+	domain, err := fork.ComputeDomain(b.beaconCfg.DomainBeaconProposer[:], utils.Uint32ToBytes4(forkVersion), headState.GenesisValidatorsRoot())
+	if err != nil {
+		return err
+	}
+	sigRoot, err := fork.ComputeSigningRoot(header.Header, domain)
+	if err != nil {
+		return err
+	}
+	pk, err := headState.ValidatorPublicKey(int(header.Header.ProposerIndex))
+	if err != nil {
+		return err
+	}
+	if ok, err = bls.Verify(header.Signature[:], sigRoot[:], pk[:]); err != nil {
+		return err
+	}
+	if !ok {
+		return fmt.Errorf("blob signature validation: signature not valid")
+	}
+	return nil
+}
+
+func (b *blobSidecarService) scheduleBlobSidecarForLaterExecution(blobSidecar *cltypes.BlobSidecar) {
+	blobSidecarJob := &blobSidecarJob{
+		blobSidecar:  blobSidecar,
+		creationTime: time.Now(),
+	}
+	b.blobSidecarsScheduledForLaterExecution.Store(blobSidecarJob, struct{}{})
```

Comment:
just left a question.
here you expect to do these job in batch, right? any idea why not use worker pool?
I am considering which is cheaper and better.

Reply:
uh yeah, i will try that. this is only one worker FYI. also not necessarily in batches